lib/pwnlib/reg_sort.rb

```diff
@@ -0,0 +1,221 @@
+# encoding: ASCII-8BIT
+
+require 'pwnlib/context'
+
+module Pwnlib
+  # Topographical sort
+  module RegSort
+    # @note Do not create and call instance method here. Instead, call module method on {RegSort}.
+    module ClassMethod
+      # Walk down the assignment list of a register,
+      # return the path walked if it is encountered again.
+      #
+      # @return [Array]
+      #   The registers that involved in the cycle start from `reg`.
+      #   Empty array will be returned if no cycle start and end at `reg`.
+      # @example
+      #   check_cycle('a', {a: 1}) #=> []
+      #   check_cycle('a', {a: 'a'}) #=> ['a']
+      #   check_cycle('a', {a: 'b', b: 'c', c: 'b', d: 'a'}) #=> []
+      #   check_cycle('a', {a: 'b', b: 'c', c: 'd', d: 'a'})
+      #   #=> ['a', 'b', 'c', 'd']
+      def check_cycle(reg, assignments)
+        check_cycle_(reg, assignments.map { |k, v| [k.to_s, v] }.to_h, [])
+      end
+
+      # Sorts register dependencies.
+      #
+      # Given a dictionary of registers to desired register contents,
+      # return the optimal order in which to set the registers to
+      # those contents.
+      #
+      # The implementation assumes that it is possible to move from
+      # any register to any other register.
+      #
+      # If a dependency cycle is encountered, one of the following will
+      # occur:
+      #
+      # - If the named register +tmp+ is set and is valid, it will be used
+      #   as a temporary register.
+      # - If +xchg+ is +true+, it is assumed that dependency cyles can
+      #   be broken by swapping the contents of two register (aka the
+      #   +xchg+ instruction on i386).
+      # - If neither +xchg+ nor +tmp+ is set, try to find a register in
+      #   +all_regs+ that is not involved in the cycle, use it as a
+      #   temporary register.
+      # - If the dependency cycle cannot be resolved as described above,
+      #   an exception is raised.
+      #
+      # @param [Hash] in_out
+      #   Dictionary of desired register states.
+      #   Keys are registers, values are either registers or any other value.
+      # @param [Array] all_regs
+      #   List of all possible registers.
+      #   Used to determine which values in +in_out+ are registers, versus
+      #   regular values.
+      # @option [String, Object] tmp
```

Comment:
`String?`, seems that we only check if `tmp` is not nil in current implementation?
Update the text below too.